File: app/src/main/java/iuliiaponomareva/eventum/MainActivity.java
package iuliiaponomareva.eventum;


import android.app.LoaderManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.Loader;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.v4.app.DialogFragment;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v4.widget.DrawerLayout;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;

public class MainActivity extends AppCompatActivity
        implements AddFeedDialogFragment.AddFeedDialogListener,
        RemoveFeedDialogFragment.RemoveFeedDialogListener,
        SwipeRefreshLayout.OnRefreshListener, LoaderManager.LoaderCallbacks<List<Channel>> {
    private static final String ADD_FEED_FRAGMENT_TAG = "addNewFeed";
    private static final String REMOVE_FEED_FRAGMENT_TAG = "removeFeed";
    private Reader reader;
    private ArrayAdapter<Channel> drawerAdapter;
    private List<News> news;
    private NewsArrayAdapter newsAdapter;
    private ListView newsListView;
    public final static String NEWS_LINK = "iuliiaponomareva.eventum.NEWS_LINK";
    private DrawerLayout drawerLayout;
    private ListView drawerListView;
    private ActionBarDrawerToggle drawerToggle;
    private Channel selectedChannel;
    private Channel all;
    private SwipeRefreshLayout refreshLayout;
    private BroadcastReceiver receiver;
    private final static int CHANNEL_LOADER_ID = 0;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        reader = new Reader();
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setHomeButtonEnabled(true);
        }
        all = new Channel(getString(R.string.all_feeds), getString(R.string.all_feeds), "", "");
        setUpChannelsView();
        setUpNewsView();
        createBroadcastReceiver();
        getLoaderManager();
    }

    private void setUpChannelsView() {
        drawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawerListView = (ListView) findViewById(R.id.drawer);
        drawerToggle = new ActionBarDrawerToggle(this, drawerLayout, R.string.drawer_open,
                R.string.drawer_close) {};
        drawerAdapter = new ArrayAdapter<>(MainActivity.this,
                R.layout.drawer_list_item, new ArrayList<Channel>());
        drawerLayout.addDrawerListener(drawerToggle);
        drawerListView.setAdapter(drawerAdapter);
        drawerListView.setOnItemClickListener(new OnItemClickListener() {

            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position,
                                    long id) {
                drawerListView.setItemChecked(position, true);
                selectedChannel = (Channel) parent.getItemAtPosition(position);
                String selectedFeed = selectedChannel.getURL();
                drawerLayout.closeDrawer(drawerListView);
                newsAdapter.cancel();
                news.clear();
                if (isConnectedToNetwork()) {
                    if (selectedChannel.equals(all)) {
                        reader.refreshAllNews(MainActivity.this);
                    }
                    else
                        reader.refreshNewsFromFeed(selectedFeed, MainActivity.this);
                } else {
                    if (selectedChannel.equals(all))
                        news.addAll(reader.getAllNews());
                    else
                        news.addAll(reader.getNewsFromFeed(selectedFeed));
                    createToast(R.string.no_internet);
                }
                setTitle(selectedChannel.getTitle());

            }
        });

    }

    private void setUpNewsView() {
        newsListView = (ListView) findViewById(R.id.all_news);
        newsListView.setOnItemClickListener(new OnItemClickListener(){
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                News selectedNews = (News) parent.getItemAtPosition(position);
                Intent intent = new Intent(MainActivity.this, ViewPageActivity.class);
                intent.putExtra(NEWS_LINK, selectedNews.getLink());
                startActivity(intent);
            }
        });
        news = new ArrayList<>();
        newsAdapter = new NewsArrayAdapter(MainActivity.this, news);
        newsListView.setAdapter(newsAdapter);
        newsListView.setEmptyView(findViewById(R.id.empty_textview));
        refreshLayout = ((SwipeRefreshLayout) findViewById(R.id.swipe_to_refresh));
        refreshLayout.setOnRefreshListener(this);
        newsListView.setOnScrollListener(new AbsListView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(AbsListView view, int scrollState) {
            }

            @Override
            public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount,
                                 int totalItemCount) {
                int topRowVerticalPosition = (newsListView == null
                        || newsListView.getChildCount() == 0) ?
                        0 : newsListView.getChildAt(0).getTop();
                refreshLayout.setEnabled((firstVisibleItem == 0) && (topRowVerticalPosition >= 0));
            }
        });
    }

    private void createBroadcastReceiver() {
        receiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                String action = intent.getAction();
                if (action.equals(ParseChannelService.ACTION_BROADCAST_CHANNELS)) {
                    Parcelable[] parcelables = intent.getParcelableArrayExtra(
                            ParseChannelService.NEW_CHANNELS);
                    Channel[] newChannels = Arrays.copyOf(parcelables, parcelables.length,
                            Channel[].class);
                    notifyAboutNewChannel(newChannels);
                } else if (action.equals(ChannelService.ACTION_CHANNELS_CHANGED)) {
                    getLoaderManager().restartLoader(CHANNEL_LOADER_ID, null, MainActivity.this);
                } else if (action.equals(NewsService.ACTION_BROADCAST_NEWS)) {
                    String[] urls = intent.getStringArrayExtra(NewsService.URLS);
                    for (String url : urls) {
                        Parcelable[] parcelables = intent.getParcelableArrayExtra(url);
                        News[] news = Arrays.copyOf(parcelables, parcelables.length, News[].class);
                        reader.finishRefreshing(news, url);
                    }
                    if (urls.length == 1)
                        showNews(reader.getNewsFromFeed(urls[0]));
                    else
                        showNews(reader.getAllNews());
                }
            }
        };
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        drawerToggle.syncState();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        drawerToggle.onConfigurationChanged(newConfig);
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_activity_action_bar, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (drawerToggle.onOptionsItemSelected(item)) {
            return true;
        }
        switch(item.getItemId()) {
            case R.id.addFeedItem:
                addFeed();
                return true;
            case R.id.removeFeedItem:
                removeFeed();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void addFeed() {
        DialogFragment newFragment = new AddFeedDialogFragment();
        newFragment.show(getSupportFragmentManager(), ADD_FEED_FRAGMENT_TAG);
    }

    private void removeFeed() {
        DialogFragment newFragment = new RemoveFeedDialogFragment();
        Bundle args = new Bundle();
        args.putStringArray(RemoveFeedDialogFragment.FEEDS, reader.getFeeds());
        newFragment.setArguments(args);
        newFragment.show(getSupportFragmentManager(), REMOVE_FEED_FRAGMENT_TAG);
    }

    @Override
    protected void onResume() {
        super.onResume();
        IntentFilter intentFilter = new IntentFilter(ParseChannelService.ACTION_BROADCAST_CHANNELS);
        intentFilter.addAction(NewsService.ACTION_BROADCAST_NEWS);
        intentFilter.addAction(ChannelService.ACTION_CHANNELS_CHANGED);
        LocalBroadcastManager.getInstance(this).registerReceiver(receiver,
                intentFilter);
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (receiver != null)
            LocalBroadcastManager.getInstance(this).unregisterReceiver(receiver);
    }

    @Override
    protected void onStop() {
        super.onStop();
        saveSelectedChannel();
    }

    private void saveSelectedChannel() {
        if (selectedChannel != null) {
            SharedPreferences sharedPref = getPreferences(Context.MODE_PRIVATE);
            SharedPreferences.Editor editor = sharedPref.edit();
            editor.putString(getString(R.string.chosen_feed_pref_label), selectedChannel.getURL());
            editor.apply();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        getLoaderManager().restartLoader(CHANNEL_LOADER_ID, null, this);
    }

    @Override
    public void addChosenFeed(String feedUrl) {
        if (isConnectedToNetwork()) {
            feedUrl = Reader.checkURL(feedUrl);
            if (reader.hasFeed(feedUrl)) {
                createToast(feedUrl + " " + getString(R.string.has_already_been_added));
            }
            else {
                reader.addFeed(this, feedUrl);
            }
        }
        else {
            createToast(R.string.no_internet);
        }
    }

    private void notifyAboutNewChannel(Channel[] newChannels) {
        if ((newChannels != null) && (newChannels.length != 0)) {
            ChannelService.startActionSave(this, newChannels);
            for (Channel channel: newChannels)
                reader.addFeed(channel);
            getLoaderManager().restartLoader(CHANNEL_LOADER_ID, null, this);
        }
        else
            createToast(R.string.error_adding_feed);
    }

    private void createToast(String message) {
        Toast toast = Toast.makeText(getApplicationContext(), message, Toast.LENGTH_LONG);
        toast.setGravity(Gravity.CENTER_HORIZONTAL, 0, 0);
        toast.show();
    }

    private void createToast(int messageRes) {
        Toast toast = Toast.makeText(getApplicationContext(), messageRes, Toast.LENGTH_LONG);
        toast.setGravity(Gravity.CENTER_HORIZONTAL, 0, 0);
        toast.show();
    }

    @Override
    public void removeChosenFeed(String feedURL) {
        if ((selectedChannel != null)&&(selectedChannel.getURL().equals(feedURL)))
            selectedChannel = all;
        drawerAdapter.remove(reader.getFeed(feedURL));
        reader.removeFeed(feedURL);
        drawerAdapter.notifyDataSetChanged();
        ChannelService.startActionDelete(this, feedURL);

    }

    private boolean isConnectedToNetwork() {
        ConnectivityManager manager = (ConnectivityManager) getApplicationContext()
                .getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = manager.getActiveNetworkInfo();
        return (networkInfo != null && networkInfo.isConnected());
    }


    @Override
    public void onRefresh() {
        refreshNews();
        refreshLayout.setRefreshing(false);
    }

    private void refreshNews() {
        if (selectedChannel == null)
            selectedChannel = all;
        if (isConnectedToNetwork()) {
            news.clear();
            if (selectedChannel.equals(all))
                reader.refreshAllNews(this);
            else
                reader.refreshNewsFromFeed(selectedChannel.getURL(), this);
            newsAdapter.notifyDataSetChanged();
        }
        else {
            createToast(R.string.no_internet);
        }
    }

    private void showNews(Set<News> newsSet) {
        newsAdapter.addAll(newsSet);
        newsAdapter.setChannelURL(selectedChannel.getURL());
        newsAdapter.notifyDataSetChanged();
        setTitle(selectedChannel.getTitle());
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        drawerLayout.removeDrawerListener(drawerToggle);
    }

    @Override
    public Loader<List<Channel>> onCreateLoader(int id, Bundle args) {
        return new ChannelsLoader(MainActivity.this, reader);
    }

    @Override
    public void onLoadFinished(Loader<List<Channel>> loader, List<Channel> data) {
        reader.addAll(data);
        drawerAdapter.clear();
        TextView textView = (TextView)findViewById(R.id.empty_textview);
        if (data.isEmpty())
            textView.setText(R.string.no_feeds_added_yet);
        else
            textView.setText(R.string.loading);
        drawerAdapter.add(all);
        drawerAdapter.addAll(data);
        drawerAdapter.notifyDataSetChanged();
        selectedChannel = reader.getFeed(getSelectedFeedFromPreferences());
        int position = drawerAdapter.getPosition(selectedChannel);
        drawerListView.setSelection(position);
        drawerListView.setItemChecked(position, true);
        if (selectedChannel == null)
            setTitle(R.string.chosen_feed_default);
        else
            setTitle(selectedChannel.getTitle());
        refreshNews();
    }

    @Override
    public void onLoaderReset(Loader<List<Channel>> loader) {

    }

    private String getSelectedFeedFromPreferences() {
        SharedPreferences sharedPref = getPreferences(Context.MODE_PRIVATE);
        String defaultValue = getString(R.string.chosen_feed_default);
        return sharedPref.getString("chosenFeed", defaultValue);
    }
}

File: app/src/main/java/iuliiaponomareva/eventum/activities/MainActivity.kt
package iuliiaponomareva.eventum.activities

import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.view.Gravity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.content.edit
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView.AdapterDataObserver
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import iuliiaponomareva.eventum.ChannelRepository
import iuliiaponomareva.eventum.R
import iuliiaponomareva.eventum.adapters.ChannelsAdapter
import iuliiaponomareva.eventum.adapters.NewsAdapter
import iuliiaponomareva.eventum.data.Channel
import iuliiaponomareva.eventum.data.News
import iuliiaponomareva.eventum.data.ReaderDatabase
import iuliiaponomareva.eventum.fragments.AddFeedDialogFragment
import iuliiaponomareva.eventum.fragments.AddFeedDialogFragment.AddFeedDialogListener
import iuliiaponomareva.eventum.fragments.RemoveFeedDialogFragment
import iuliiaponomareva.eventum.fragments.RemoveFeedDialogFragment.RemoveFeedDialogListener
import iuliiaponomareva.eventum.viewModels.ChannelError
import iuliiaponomareva.eventum.viewModels.ChannelViewModel
import iuliiaponomareva.eventum.viewModels.ChannelViewModelFactory
import iuliiaponomareva.eventum.viewModels.NewsViewModel
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

class MainActivity : AppCompatActivity(), AddFeedDialogListener,
    RemoveFeedDialogListener, OnRefreshListener {
    private lateinit var drawerAdapter: ChannelsAdapter
    private lateinit var newsAdapter: NewsAdapter
    private lateinit var adapterDataObserver: AdapterDataObserver
    private lateinit var drawerToggle: ActionBarDrawerToggle
    private lateinit var selectedChannel: Channel
    private lateinit var all: Channel
    private lateinit var channelsViewModel: ChannelViewModel
    private lateinit var newsViewModel: NewsViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeButtonEnabled(true)
        all = Channel(
            getString(R.string.all_feeds),
            getString(R.string.all_feeds),
            "",
            ""
        )
        setUpChannelsView()
        setUpNewsView()
        val channelDao = ReaderDatabase.getDatabase(applicationContext).channelDao()
        val repository = ChannelRepository(channelDao)
        channelsViewModel = ViewModelProvider(this,
            ChannelViewModelFactory(
                repository
            )
        )
            .get(ChannelViewModel::class.java)
        newsViewModel = ViewModelProvider(this)[NewsViewModel::class.java]
        channelsViewModel.channels.observe(this, { event ->
            val handled = event.handled
            onChannelsLoaded(event.getEvent(), refreshNews = !handled)
        })
        channelsViewModel.error.observe(this, {
            when (val error = it.getEventIfNotHandled()) {
                ChannelError.ALREADY_EXISTS -> createToast(error.url + " " + getString(R.string.has_already_been_added))
                ChannelError.ERROR_ADDING -> createToast(R.string.error_adding_feed)
            }
        })

        newsViewModel.allNews.observe(this, { newsMap ->
            if (selectedChannel == all) {
                showNews(newsMap.values.flatten())
            } else {
                showNews(newsMap[selectedChannel.url] ?: listOf())
            }
        })
    }

    private fun setUpChannelsView() {
        drawerToggle = ActionBarDrawerToggle(
            this, drawerLayout, R.string.drawer_open,
            R.string.drawer_close
        )
        drawerAdapter = ChannelsAdapter(onChannelSelected())
        drawer.adapter = drawerAdapter
        drawerLayout.addDrawerListener(drawerToggle)
    }

    private fun onChannelSelected(): (Channel) -> Unit {
        return fun(channel: Channel) {
            selectedChannel = channel
            drawerLayout.closeDrawer(drawer)
            supportActionBar?.title = selectedChannel.title
            newsAdapter.cancel()
            if (isConnectedToNetwork()) {
                if (selectedChannel == all) {
                    val urls =
                        channelsViewModel.channels.value?.getEvent()?.map { it.url }?.toTypedArray()
                            ?: arrayOf()
                    newsViewModel.refreshNews(urls)
                } else {
                    newsViewModel.refreshNews(arrayOf(selectedChannel.url))
                }
            } else {
                val allNews = newsViewModel.allNews.value
                if (selectedChannel == all) {
                    showNews(allNews?.values?.flatten() ?: listOf())
                } else {
                    showNews(allNews?.get(selectedChannel.url) ?: listOf())
                }
                createToast(R.string.no_internet)
            }
        }
    }

    private fun setUpNewsView() {
        refreshLayout.setOnRefreshListener(this)
        newsAdapter = NewsAdapter { news ->
            if (!TextUtils.isEmpty(news.link)) {
                val intent = Intent(this@MainActivity, ViewPageActivity::class.java)
                intent.putExtra(NEWS_LINK, news.link)
                startActivity(intent)
            }
        }
        newsRecyclerView.adapter = newsAdapter
        val dividerItemDecoration = DividerItemDecoration(this, LinearLayoutManager.VERTICAL)
        val drawable = ContextCompat.getDrawable(this, R.drawable.divider)
        drawable?.let { dividerItemDecoration.setDrawable(it) }
        newsRecyclerView.addItemDecoration(dividerItemDecoration)
        newsRecyclerView.setHasFixedSize(true)
        adapterDataObserver = object: AdapterDataObserver() {
            override fun onChanged() {
                super.onChanged()
                if (newsAdapter.itemCount == 0) {
                    emptyView.visibility = View.VISIBLE
                } else {
                    emptyView.visibility = View.INVISIBLE
                }
            }
        }
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        drawerToggle.syncState()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        drawerToggle.onConfigurationChanged(newConfig)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_activity_action_bar, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (drawerToggle.onOptionsItemSelected(item)) {
            true
        } else when (item.itemId) {
            R.id.addFeedItem -> {
                addFeed()
                true
            }
            R.id.removeFeedItem -> {
                removeFeed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun addFeed() {
        val newFragment: DialogFragment = AddFeedDialogFragment()
        newFragment.show(supportFragmentManager, ADD_FEED_FRAGMENT_TAG)
    }

    private fun removeFeed() {
        val newFragment: DialogFragment = RemoveFeedDialogFragment()
        val channels = ArrayList<Channel>()
        val event = channelsViewModel.channels.value?.getEvent() ?: listOf()
        channels.addAll(event)
        val args = Bundle()
        args.putParcelableArrayList(RemoveFeedDialogFragment.FEEDS, channels)
        newFragment.arguments = args
        newFragment.show(
            supportFragmentManager,
            REMOVE_FEED_FRAGMENT_TAG
        )
    }

    override fun onStart() {
        super.onStart()
        newsAdapter.registerAdapterDataObserver(adapterDataObserver)
    }

    override fun onStop() {
        super.onStop()
        newsAdapter.unregisterAdapterDataObserver(adapterDataObserver)
        saveSelectedChannel()
    }

    private fun saveSelectedChannel() {
        val sharedPref = getPreferences(Context.MODE_PRIVATE)
        sharedPref.edit {
            putString(getString(R.string.chosen_feed_pref_label), selectedChannel.url)
        }
    }

    override fun addChosenFeed(feedUrl: String) {
        if (isConnectedToNetwork()) {
            channelsViewModel.addChannel(feedUrl)
        } else {
            createToast(R.string.no_internet)
        }
    }

    private fun createToast(message: String) {
        val toast = Toast.makeText(applicationContext, message, Toast.LENGTH_LONG)
        toast.setGravity(Gravity.CENTER_HORIZONTAL, 0, 0)
        toast.show()
    }

    private fun createToast(messageRes: Int) {
        val toast = Toast.makeText(applicationContext, messageRes, Toast.LENGTH_LONG)
        toast.setGravity(Gravity.CENTER_HORIZONTAL, 0, 0)
        toast.show()
    }

    override fun removeChosenFeed(feed: Channel) {
        if (selectedChannel == feed) {
            selectedChannel = all
        }
        drawerAdapter.remove(feed)
        channelsViewModel.deleteChannel(feed)
    }

    @Suppress("DEPRECATION")
    private fun isConnectedToNetwork(): Boolean {
        val manager = applicationContext
            .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val network = manager.activeNetwork ?: return false
            val networkCapabilities = manager.getNetworkCapabilities(network) ?: return false
            return networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
        } else {
            val networkInfo = manager.activeNetworkInfo
            return networkInfo != null && networkInfo.isConnected
        }
    }

    override fun onRefresh() {
        refreshNews()
        refreshLayout.isRefreshing = false
    }

    private fun refreshNews() {
        if (!this::selectedChannel.isInitialized) {
            selectedChannel = all
        }
        if (isConnectedToNetwork()) {
            newsAdapter.news = listOf()
            if (selectedChannel == all) {
                val urls = channelsViewModel.channels.value?.getEvent()?.map { it.url }?.toTypedArray()
                    ?: arrayOf()
                newsViewModel.refreshNews(urls)
            } else {
                newsViewModel.refreshNews(arrayOf(selectedChannel.url))
            }
        } else {
            createToast(R.string.no_internet)
        }
    }

    private fun showNews(newsCollection: Collection<News>) {
        newsAdapter.channelURL = selectedChannel.url
        newsAdapter.news = newsCollection.toList()
    }

    override fun onDestroy() {
        super.onDestroy()
        drawerLayout.removeDrawerListener(drawerToggle)
    }


    private fun onChannelsLoaded(data: List<Channel>, refreshNews: Boolean = true) {
        drawerAdapter.clear()
        if (data.isEmpty()) {
            emptyView.setText(R.string.no_feeds_added_yet)
        } else {
            emptyView.setText(R.string.loading)
        }
        val channels = mutableListOf(all)
        channels.addAll(data)
        drawerAdapter.addAll(channels)
        if (!this::selectedChannel.isInitialized) {
            selectedChannel = data.find { it.url == selectedFeedFromPreferences } ?: all
        }
        supportActionBar?.title = selectedChannel.title
        if (refreshNews) {
            refreshNews()
        } else {
            if (selectedChannel == all) {
                showNews(newsViewModel.allNews.value?.values?.flatten() ?: listOf())
            } else {
                showNews(newsViewModel.allNews.value?.get(selectedChannel.url) ?: listOf())
            }
        }
    }

    private val selectedFeedFromPreferences: String
        get() {
            val sharedPref = getPreferences(Context.MODE_PRIVATE)
            val defaultValue = getString(R.string.chosen_feed_default)
            return sharedPref.getString("chosenFeed", defaultValue)!!
        }

    companion object {
        private const val ADD_FEED_FRAGMENT_TAG = "addNewFeed"
        private const val REMOVE_FEED_FRAGMENT_TAG = "removeFeed"
        const val NEWS_LINK = "iuliiaponomareva.eventum.NEWS_LINK"
    }
}